### PR TITLE
State cache

### PR DESCRIPTION
**Pull request by canepat**
This PR introduces a coherent state cache working on top of a remote read-only transaction provided by the KV interface.

The pair state cache + read-only transaction must provide "Serializable Isolation Level" semantic: all data form a consistent database view at the moment when the read-only transaction started, read data are immutable until the end of the read-only transaction, the reader can't see newer updates.

### CODE REVIEW

File: silkrpc/ethdb/kv/state_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    void evict_roots();
+
+    CoherentCacheConfig config_;
+
+    std::map<StateViewId, std::unique_ptr<CoherentStateRoot>> state_view_roots_;
+    StateViewId latest_state_view_id_{0};
+    CoherentStateRoot* latest_state_view_{nullptr};
+    std::list<KeyValue> state_evictions_;
+    std::list<KeyValue> code_evictions_;
+    std::shared_mutex rw_mutex_;
```
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.hpp:141]: What about opening an issue on this topic and let this PR as it is?

File: silkrpc/ethdb/kv/state_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+    state_eviction_count_ = state_evictions_.size();
+    code_eviction_count_ = code_evictions_.size();
+
+    return root;
+}
+
+void CoherentStateCache::evict_roots() {
+    SILKRPC_DEBUG << "CoherentStateCache::evict_roots latest_state_view_id_=" << latest_state_view_id_ << "\n";
+    if (latest_state_view_id_ <= config_.max_views) {
```
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp:348]: :+1: Thanks for spotting this!

File: silkrpc/ethdb/kv/state_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    }
+    const auto [new_root_it, _] = state_view_roots_.emplace(view_id, std::make_unique<CoherentStateRoot>());
+    SILKRPC_DEBUG << "CoherentStateCache::get_root view_id=" << view_id << " root=" << root_it->second.get() << " created\n";
+    return new_root_it->second.get();
+}
+
+CoherentStateRoot* CoherentStateCache::advance_root(StateViewId view_id) {
+    CoherentStateRoot* root = get_root(view_id);
+
+    const auto previous_root_it = state_view_roots_.find(view_id - 1);
```
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp:317]: See response above

File: silkrpc/ethdb/kv/state_cache.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp]: After a closer look, I've realized that such check is buggy but tries to address a real corner case, albeit far away in future: view ID wrapping.

IMO in such case it's better just clearing the cache to keep things as simple as possible.

File: silkrpc/ethdb/kv/state_cache.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/ethdb/kv/state_cache.hpp:150]: ok

File: silkrpc/ethdb/kv/state_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    if (next_view_id == 0) {
+        // Next view ID is zero with cache not empty => view ID wrapping => clear the cache except for new latest view
+        std::erase_if(state_view_roots_, [&](const auto& item) {
+            auto const& [view_id, _] = item;
+            return view_id != next_view_id;
+        });
+        return;
+    }
+    // Erase older state views in order not to exceed max_views
+    const auto max_view_id_to_delete = latest_state_view_id_ - config_.max_views + 1;
```
Comment (lupin012) [inline silkrpc/ethdb/kv/state_cache.cpp:360]: if we use: max_view_id_to_delete = latest_state_view_id_ - config_.max_views + 1;
the view_id < max_view_id_to_delete;
or 
if we use: max_view_id_to_delete = latest_state_view_id_ - config_.max_views;
the view_id <= max_view_id_to_delete;

File: silkrpc/ethdb/kv/state_cache.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lupin012) [reply silkrpc/ethdb/kv/state_cache.cpp]: ok

File: silkrpc/ethdb/kv/state_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    co_return co_await cache_->get_code(key, txn_);
+}
+
+CoherentStateCache::CoherentStateCache(CoherentCacheConfig config) : config_(config) {
+    if (config.max_views == 0) {
+        throw std::invalid_argument{"unexpected zero max_views"};
+    }
+}
+
+std::unique_ptr<StateView> CoherentStateCache::get_view(Transaction& txn) {
```
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp:50]: This `silkrpc::ethdb::Transaction` is already distinct from `silkrpc::Transaction` thanks to different namespaces.

File: silkrpc/ethdb/kv/state_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    return latest_state_view_->code_cache.size();
+}
+
+void CoherentStateCache::on_new_block(const remote::StateChangeBatch& state_changes) {
+    if (state_changes.changebatch_size() == 0) {
+        SILKRPC_WARN << "Unexpected empty batch received and skipped\n";
+        return;
+    }
+
+    std::unique_lock write_lock{rw_mutex_};
```
[the remark this answers is not in the document]
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp:78]: I would left as it is just for readability purposes (the difference, if any, is negligible)

File: silkrpc/ethdb/kv/state_cache.cpp (hunk 4 above)
Comment (canepat) [reply silkrpc/ethdb/kv/state_cache.cpp:360]: This is not the case because `evict_roots` is called before `latest_view_id_` gets updated

**Review by lupin012 — APPROVED** [review]

**Pull request merged**